File: packages/asdl-slots/src/asdl_slots/cli/slot/gc.py
"""`slot gc` — free slots whose branch has a merged or closed PR."""

from __future__ import annotations

import sys
from typing import Annotated

import click
from rich.console import Console

from asdl_core import get_console
from asdl_core.clinkr.ensure import Ensure
from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from asdl_core.gh.types import PRState
from asdl_slots.cli.slot.cleanup_rendering import (
    SlotCleanupResult,
    cleanup_failure_text,
    cleanup_preview_line,
    cleanup_result_line,
    cleanup_skipped_text,
    cleanup_success_text,
    cleanup_to_result,
)
from asdl_slots.cli.slot.context import load_slots_context
from asdl_slots.lifecycle.outcomes import (
    SlotFreeCleanupAction,
    SlotGcAction,
    SlotGcOutcome,
    SlotLifecycleFailure,
)
from asdl_slots.lifecycle.release import (
    execute_gc_release,
    outcome_from_gc_release_preview,
    plan_gc_release,
)
from asdl_slots.repo_context import NoRepoSentinel


class SlotGcRequest(ClinkrModel):
    dry_run: Annotated[bool, click.Option(["--dry-run"], is_flag=True, default=False)] = False
    force: Annotated[bool, click.Option(["-f", "--force"], is_flag=True, default=False)] = False
    delete_branches: Annotated[
        bool,
        click.Option(
            ["--delete-branches"],
            is_flag=True,
            default=False,
            help="Force-delete local branches after successfully freeing eligible slots.",
        ),
    ] = False


SLOT_GC_DELETE_BRANCH_CLEANUP_ACTIONS: tuple[SlotFreeCleanupAction, ...] = ("local_branch",)


class SlotGcResultEntry(ClinkrModel):
    slot_name: str
    branch_name: str
    worktree_path: str
    action: SlotGcAction
    pr_number: int | None
    pr_state: PRState | None
    pr_url: str | None
    message: str | None
    cleanup: tuple[SlotCleanupResult, ...] = ()


class SlotGcResult(ClinkrModel):
    entries: tuple[SlotGcResultEntry, ...]
    freed_count: int
    kept_count: int
    skipped_count: int
    error_count: int
    cleanup_error_count: int = 0
    dry_run: bool
    cancelled: bool = False


_ACTION_LABELS: dict[SlotGcAction, tuple[str, str]] = {
    "freed": ("[green]✓ freed[/green]", "green"),
    "would_free": ("[yellow]→ would free[/yellow]", "yellow"),
    "kept_open_pr": ("[blue]• kept (open PR)[/blue]", "blue"),
    "kept_no_pr": ("[dim]• kept (no PR)[/dim]", "dim"),
    "skipped_dirty": ("[yellow]! skipped (dirty)[/yellow]", "yellow"),
    "skipped_operation": ("[yellow]! skipped (operation)[/yellow]", "yellow"),
    "error": ("[red]✗ error[/red]", "red"),
}


def _gc_console(*, err: bool = False) -> Console:
    if err:
        return Console(file=sys.stderr)
    return get_console()


def render_slot_gc(result: SlotGcResult, *, err: bool = False) -> None:
    console = _gc_console(err=err)
    if result.cancelled:
        console.print("[yellow]Cancelled — no slots freed.[/yellow]")
        return
    if not result.entries:
        console.print("[dim]No assignments to sweep.[/dim]")
        return

    for entry in result.entries:
        label, _colour = _ACTION_LABELS[entry.action]
        pr_suffix = ""
        if entry.pr_number is not None:
            pr_suffix = f" [dim]PR #{entry.pr_number} {entry.pr_state}[/dim]"
        console.print(
            f"{label} [bold cyan]{entry.slot_name}[/bold cyan] "
            f"([yellow]{entry.branch_name}[/yellow]){pr_suffix}"
        )
        if entry.message:
            console.print(f"    [dim]{entry.message}[/dim]")
        for cleanup in entry.cleanup:
            if result.dry_run:
                console.print(f"    {cleanup_preview_line(cleanup)}")
            else:
                console.print(f"    {cleanup_result_line(cleanup)}")

    verb = "Would free" if result.dry_run else "Freed"
    summary = (
        f"\n[bold]{verb} {result.freed_count}[/bold]; "
        f"kept {result.kept_count}; "
        f"skipped {result.skipped_count}; "
        f"errors {result.error_count}"
    )
    if result.cleanup_error_count:
        summary = f"{summary}; cleanup errors {result.cleanup_error_count}"
    console.print(summary)


def _confirm_free_slots(count: int, *, delete_branches: bool = False) -> bool:
    """Confirm on stderr without leaking echoed test input into stdout."""
    while True:
        prompt = (
            f"Free {count} slot(s) and delete local branches? [Y/n]: "
            if delete_branches
            else f"Free {count} slot(s)? [Y/n]: "
        )
        click.echo(prompt, err=True, nl=False)
        sys.stderr.flush()
        raw_value = sys.stdin.readline()
        if raw_value == "":
            raise click.Abort()
        value = raw_value.strip().lower()
        if value in ("", "y", "yes"):
            return True
        if value in ("n", "no"):
            return False
        click.echo("Error: invalid input", err=True)


def _result_from_outcome(outcome: SlotGcOutcome, *, cancelled: bool = False) -> SlotGcResult:
    return SlotGcResult(
        entries=tuple(
            SlotGcResultEntry(
                slot_name=e.slot_name,
                branch_name=e.branch_name,
                worktree_path=str(e.worktree_path),
                action=e.action,
                pr_number=e.pr_number,
                pr_state=e.pr_state,
                pr_url=e.pr_url,
                message=e.message,
                cleanup=cleanup_to_result(e.cleanup),
            )
            for e in outcome.entries
        ),
        freed_count=0 if cancelled else outcome.freed_count,
        kept_count=outcome.kept_count,
        skipped_count=outcome.skipped_count,
        error_count=outcome.error_count,
        cleanup_error_count=outcome.cleanup_error_count,
        dry_run=outcome.dry_run,
        cancelled=cancelled,
    )


def _cleanup_error_message(result: SlotGcResult) -> str:
    lines: list[str] = []
    for entry in result.entries:
        if not any(cleanup.status == "error" for cleanup in entry.cleanup):
            continue
        if result.dry_run:
            lines.append(f"Would free {entry.slot_name} ({entry.branch_name})")
        else:
            lines.append(f"✓ Freed {entry.slot_name} ({entry.branch_name})")
            lines.append(f"  Worktree kept at {entry.worktree_path}; detached HEAD at trunk")
        for cleanup in entry.cleanup:
            if cleanup.status == "success":
                lines.append(f"  ✓ {cleanup_success_text(cleanup)}")
            elif cleanup.status == "skipped":
                lines.append(f"  - {cleanup_skipped_text(cleanup)}")
            elif cleanup.status == "error":
                lines.append(f"  ✗ {cleanup_failure_text(cleanup)}")
    if not lines:
        return "Cleanup failed."
    return "\n".join(lines)


def _exit_for_result(result: SlotGcResult) -> ClinkrExit[SlotGcResult]:
    if result.cleanup_error_count:
        return ClinkrExit.negative(result, message=_cleanup_error_message(result))
    return ClinkrExit.ok(result)


@clinkr_operation(
    name="gc",
    help="Free slots whose branch has a merged or closed PR.",
    human_renderer=render_slot_gc,
)
def run_slot_gc(ctx: click.Context, request: SlotGcRequest) -> ClinkrExit[SlotGcResult]:
    slots_ctx = load_slots_context(ctx)
    if isinstance(slots_ctx, NoRepoSentinel):
        Ensure.fail(error_type="not_in_repo", message=slots_ctx.message)

    Ensure.true(
        not (request.dry_run and request.force),
        error_type="conflicting_flags",
        message="--dry-run and --force are mutually exclusive.",
    )

    cleanup_actions = SLOT_GC_DELETE_BRANCH_CLEANUP_ACTIONS if request.delete_branches else ()

    preview = plan_gc_release(slots_ctx, cleanup_actions=cleanup_actions)
    if isinstance(preview, SlotLifecycleFailure):
        return ClinkrExit.failure(error_type=preview.error_type, message=preview.message)

    if request.dry_run:
        result = _result_from_outcome(preview.outcome)
        return _exit_for_result(result)

    if preview.plan.would_free_count == 0:
        outcome = outcome_from_gc_release_preview(preview, dry_run=False)
        return ClinkrExit.ok(_result_from_outcome(outcome))

    if request.force:
        result = _result_from_outcome(execute_gc_release(slots_ctx, preview))
        return _exit_for_result(result)

    render_slot_gc(_result_from_outcome(preview.outcome), err=True)
    sys.stderr.flush()
    proceed = _confirm_free_slots(
        preview.plan.would_free_count,
        delete_branches=request.delete_branches,
    )
    if proceed:
        result = _result_from_outcome(execute_gc_release(slots_ctx, preview))
        return _exit_for_result(result)
    return ClinkrExit.ok(
        _result_from_outcome(
            outcome_from_gc_release_preview(preview, dry_run=False),
            cancelled=True,
        )
    )

File: ts/packages/slot/src/operations/gc.ts
import { confirmFromStdin, failure, negative, ok } from "@asdl/clinkr";
import { z } from "zod";

import type { RepoSlotContext, SlotCliContext } from "../context.ts";
import { outcomeFromGcPlan, planGc, planGcCleanup, executeGcPlan, type SlotGcOutcome } from "../lifecycle/gc.ts";
import type { SlotFreeCleanupAction } from "../lifecycle/release-cleanup.ts";
import { cleanupPreviewLine, cleanupResultLine } from "./cleanup-rendering.ts";
import { cleanupSchema } from "./result-schemas.ts";
const gcEntrySchema = z.object({
	slot_name: z.string(),
	branch_name: z.string(),
	worktree_path: z.string(),
	action: z.union([z.literal("freed"), z.literal("would_free"), z.literal("kept_open_pr"), z.literal("kept_no_pr"), z.literal("skipped_dirty"), z.literal("skipped_operation"), z.literal("error")]),
	pr_number: z.number().int().nullable(),
	pr_state: z.union([z.literal("OPEN"), z.literal("CLOSED"), z.literal("MERGED")]).nullable(),
	pr_url: z.string().nullable(),
	message: z.string().nullable(),
	cleanup: z.array(cleanupSchema),
});

export const gcRequestSchema = z.object({
	dry_run: z.boolean().default(false).describe("Preview without mutating."),
	force: z.boolean().default(false).describe("Free candidates without prompting."),
	delete_branches: z.boolean().default(false).describe("Delete local branches for freed slots."),
});

export const gcResultSchema = z.object({
	entries: z.array(gcEntrySchema),
	freed_count: z.number().int(),
	kept_count: z.number().int(),
	skipped_count: z.number().int(),
	error_count: z.number().int(),
	dry_run: z.boolean(),
	cleanup_error_count: z.number().int(),
	cancelled: z.boolean().optional(),
});

export type GcRequest = z.infer<typeof gcRequestSchema>;
export type GcResult = z.infer<typeof gcResultSchema>;

export async function runGc(ctx: SlotCliContext, request: GcRequest) {
	if (ctx.repo.type !== "repo") return failure(ctx.repo.errorType, ctx.repo.message);
	if (request.dry_run && request.force) return failure("conflicting_flags", "--dry-run and --force cannot be combined.");
	const repoCtx: RepoSlotContext = { ...ctx, repo: ctx.repo };
	const cleanupActions: readonly SlotFreeCleanupAction[] = request.delete_branches ? ["local_branch"] : [];
	const plan = await planGc(repoCtx);
	if (plan.type === "failure") return failure(plan.failure.error_type, plan.failure.message);
	if (request.dry_run) {
		const cleanup = await planGcCleanup(repoCtx, plan.outcome, cleanupActions);
		return ok(toGcResult(outcomeFromGcPlan(plan.outcome, { isDryRun: true, cleanup })));
	}
	if (plan.outcome.would_free_count === 0) return ok(toGcResult(outcomeFromGcPlan(plan.outcome, { isDryRun: false })));
	if (!request.force) {
		if (!ctx.shouldWriteCdDirective) return failure("confirmation_required", "Destructive gc requires --force in JSON mode (or use --dry-run first).");
		const cleanup = await planGcCleanup(repoCtx, plan.outcome, cleanupActions);
		repoCtx.stderr(`${renderGc(toGcResult(outcomeFromGcPlan(plan.outcome, { isDryRun: true, cleanup })))}\n`);
		const accepted = await confirmFromStdin({ stdin: repoCtx.stdin, stderr: repoCtx.stderr, prompt: confirmationPrompt(plan.outcome.would_free_count, { shouldDeleteBranches: request.delete_branches }), defaultAnswer: "yes" });
		if (typeof accepted !== "string") return accepted;
		if (accepted === "no") return ok(toGcResult(outcomeFromGcPlan(plan.outcome, { isDryRun: false, cleanup }), { isCancelled: true }));
	}
	const outcome = await executeGcPlan(repoCtx, plan.outcome, { cleanupActions });
	const result = toGcResult(outcome);
	if (outcome.cleanup_error_count > 0) return negative("Slot gc completed with cleanup errors.", result);
	return ok(result);
}

export function renderGc(result: GcResult): string {
	if (result.cancelled === true) return ansi("yellow", "Cancelled — no slots freed.");
	if (result.entries.length === 0) return ansi("dim", "No assignments to sweep.");
	const lines: string[] = [];
	for (const entry of result.entries) {
		const pr = entry.pr_number === null ? "" : ` ${ansi("dim", `PR #${entry.pr_number} ${entry.pr_state}`)}`;
		lines.push(`${actionLabel(entry.action)} ${ansi("boldCyan", entry.slot_name)} (${ansi("yellow", entry.branch_name)})${pr}`);
		if (entry.message !== null) lines.push(`    ${ansi("dim", entry.message)}`);
		for (const cleanup of entry.cleanup) lines.push(`    ${result.dry_run ? cleanupPreviewLine(cleanup) : cleanupResultLine(cleanup)}`);
	}
	const verb = result.dry_run ? "Would free" : "Freed";
	let summary = `\n${ansi("bold", `${verb} ${result.freed_count}`)}; kept ${result.kept_count}; skipped ${result.skipped_count}; errors ${result.error_count}`;
	if (result.cleanup_error_count > 0) summary = `${summary}; cleanup errors ${result.cleanup_error_count}`;
	lines.push(summary);
	return lines.join("\n");
}

function actionLabel(action: GcResult["entries"][number]["action"]): string {
	if (action === "freed") return ansi("green", "✓ freed");
	if (action === "would_free") return ansi("yellow", "→ would free");
	if (action === "kept_open_pr") return ansi("blue", "• kept (open PR)");
	if (action === "kept_no_pr") return ansi("dim", "• kept (no PR)");
	if (action === "skipped_dirty") return ansi("yellow", "! skipped (dirty)");
	if (action === "skipped_operation") return ansi("yellow", "! skipped (operation)");
	return ansi("red", "✗ error");
}

type AnsiStyle = "bold" | "dim" | "red" | "green" | "yellow" | "blue" | "boldCyan";

const ANSI_CODES = {
	bold: "1",
	dim: "2",
	red: "31",
	green: "32",
	yellow: "33",
	blue: "34",
	boldCyan: "1;36",
} as const satisfies Record<AnsiStyle, string>;

function ansi(style: AnsiStyle, text: string): string {
	return `\u001b[${ANSI_CODES[style]}m${text}\u001b[0m`;
}

function confirmationPrompt(count: number, options: { shouldDeleteBranches: boolean }): string {
	if (options.shouldDeleteBranches) return `Free ${count} slot(s) and delete local branches? [Y/n]: `;
	return `Free ${count} slot(s)? [Y/n]: `;
}

function toGcResult(outcome: SlotGcOutcome, options: { isCancelled?: boolean | undefined } = {}): GcResult {
	return {
		entries: outcome.entries.map((entry) => ({ ...entry, cleanup: [...entry.cleanup] })),
		freed_count: options.isCancelled === true ? 0 : outcome.freed_count,
		kept_count: outcome.kept_count,
		skipped_count: outcome.skipped_count,
		error_count: outcome.error_count,
		dry_run: outcome.dry_run,
		cleanup_error_count: outcome.cleanup_error_count,
		...(options.isCancelled === undefined ? {} : { cancelled: options.isCancelled }),
	};
}

